source/main.cc

```diff
@@ -412,92 +434,68 @@ int main (int argc, char *argv[])
-#else
-          argc != 2
-#endif
-        )
-        // or if the user called aspect with either --help or -h as
-        // the sole arguments
-        ||
-        (
-          (
-#ifdef ASPECT_USE_PETSC
-            argc >= 2
-#else
-            argc == 2
-#endif
-          )
-          &&
-          (
-            (std::string(argv[1]) == "--help")
-            ||
-            (std::string(argv[1]) == "-h")
-          )
-        )
-      )
+      while (current_idx<argc)
         {
-          // print usage info only on processor 0
-          if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+          std::string arg = std::string(argv[current_idx++]);
+          if (arg == "--print-xml-schema")
             {
-              std::cout << "Usage: ./aspect <parameter_file.prm>   (to read from an input file)"
-                        << std::endl
-                        << "    or ./aspect --                     (to read from stdin)"
-                        << std::endl
-                        << std::endl;
-              if (argc >= 2)
-                std::cout << "       ./aspect --version              (for information about library versions)"
-                          << std::endl
-                          << "       ./aspect --help                 (for this usage help)"
-                          << std::endl
-                          << std::endl;
-
+              print_xml = true;
+            }
+          else if (arg=="-h" || arg =="--help")
+            {
+              if (i_am_proc_0)
+                {
+                  print_aspect_header(std::cout);
+                  print_help();
+                }
+              return 0;
+            }
+          else if (arg=="-v" || arg =="--version")
+            {
+              if (i_am_proc_0)
+                {
+                  print_aspect_header(std::cout);
+                  print_version_information(std::cout);
+                }
+              return 0;
             }
-
-          if (argc >= 2)
-            return 0;    // return without error if called with --help
           else
-            return 2;    // return error if no argument is given at all
+            {
+              prm_name = arg;
+              break;
+            }
         }
 
-      // check whether ASPECT was called with --version or -v.
-      //
-      // for the same reason as above, ignore further arguments if we
-      // are using PETSc
-      if ((
-#ifdef ASPECT_USE_PETSC
-            argc >= 2
-#else
-            argc == 2
+
+      // if no parameter given or somebody gave additional parameters,
+      // show help and exit.
+      // however, this does not work with PETSc because for PETSc, one
+      // may pass any number of flags on the command line; unfortunately,
+      // the PETSc initialization code (run through the call to
+      // MPI_InitFinalize above) does not filter these out.
+      if ((prm_name == "")
+#ifndef ASPECT_USE_PETSC
+          || (current_idx < argc)
```

Comment:
The context diff is hard to read, so maybe I'm missing something, but how can this happen? Don't we read all arguments?